Doc/howto/a-conceptual-overview-of-asyncio.rst

```diff
@@ -0,0 +1,501 @@
+           "Beginning asynchronous sleep at time: "
+           f"{datetime.datetime.now().strftime("%H:%M:%S")}."
+       )
+       await asyncio.create_task(async_sleep(3))
+       print(
+           "Done asynchronous sleep at time: "
+           f"{datetime.datetime.now().strftime("%H:%M:%S")}."
+       )
+       # asyncio.gather effectively awaits each task in the collection.
+       await asyncio.gather(*work_tasks)
+
+
+Below, we use a future to enable custom control over when that task will be
+marked as done.
+If :meth:`future.set_result() <asyncio.Future.set_result>` (the method
+responsible for marking that future as done) is never called, then this task
+will never finish.
+We've also enlisted the help of another task, which we'll see in a moment, that
+will monitor how much time has elapsed and accordingly call
+``future.set_result()``.
+
+::
+
+   async def async_sleep(seconds: float):
+       future = asyncio.Future()
+       time_to_wake = time.time() + seconds
+       # Add the watcher-task to the event loop.
+       watcher_task = asyncio.create_task(_sleep_watcher(future, time_to_wake))
+       # Block until the future is marked as done.
+       await future
+
+We'll use a rather bare object, ``YieldToEventLoop()``, to ``yield`` from
+``__await__`` in order to cede control to the event loop.
+This is effectively the same as calling ``asyncio.sleep(0)``, but this approach
+offers more clarity, not to mention it's somewhat cheating to use
+``asyncio.sleep`` when showcasing how to implement it!
+
+As usual, the event loop cycles through its queue of tasks, giving them control
+and receiving control back when they pause or finish.
+The ``watcher_task``, which runs the coroutine ``_sleep_watcher(...)``, will
+be invoked once per full cycle of the event loop's queue.
+On each resumption, it'll check the time and if not enough has elapsed, then
+it'll pause once again and hand control back to the event loop.
+Eventually, enough time will have elapsed, and ``_sleep_watcher(...)`` will
+mark the future as done, and then itself finish too by breaking out of the
+infinite ``while`` loop.
+Given this helper task is only invoked once per cycle of the event loop's queue,
+you'd be correct to note that this asynchronous sleep will sleep *at least*
+three seconds, rather than exactly three seconds.
+Note this is also of true of ``asyncio.sleep``.
+
+::
+
+   class YieldToEventLoop:
+       def __await__(self):
+           yield
+
+   async def _sleep_watcher(future, time_to_wake):
+       while True:
+           if time.time() >= time_to_wake:
+               # This marks the future as done.
+               future.set_result(None)
+               break
+           else:
+               await YieldToEventLoop()
+
+Here is the full program's output:
+
+.. code-block:: none
+
+   $ python custom-async-sleep.py
+   Beginning asynchronous sleep at time: 14:52:22.
+   I am worker. Work work.
+   I am worker. Work work.
+   I am worker. Work work.
```

Comment:
```suggestion
   I am a worker. Work work.
   I am a worker. Work work.
   I am a worker. Work work.
```

Reply:
I think this has been resolved per the discussion in another comment thread.